dl/src/main/scala/com/intel/analytics/bigdl/dataset/Transformer.scala

```diff
@@ -72,63 +73,109 @@ class Identity[A] extends Transformer[A, A] {
       private var featureData: Array[T] = null
       private var labelData: Array[T] = null
       private val batchSize = Utils.getBatchSize(totalBatch)
       private var featureSize: Array[Int] = null
       private var labelSize: Array[Int] = null
-      private var _featureLength: Int = 0
-      private var _labelLength: Int = 0
+      private var oneFeatureLength: Int = 0
+      private var oneLabelLength: Int = 0
+      private var featureArrayOfTensor: Array[Tensor[T]] = null
+      private var labelArrayOfTensor: Array[Tensor[T]] = null
       override def hasNext: Boolean = prev.hasNext
 
+      if (isTable) {
+        featureTable = T()
+        labelTable = T()
+        featureArrayOfTensor = new Array[Tensor[T]](batchSize)
+        labelArrayOfTensor = new Array[Tensor[T]](batchSize)
+        var i = 0
+        while (i < batchSize) {
+          featureArrayOfTensor(i) = Tensor[T]()
+          labelArrayOfTensor(i) = Tensor[T]()
+          i += 1
+        }
+      } else {
+        featureTensor = Tensor[T]()
+        labelTensor = Tensor[T]()
+      }
+
       override def next(): MiniBatch[T] = {
         if (prev.hasNext) {
-          var i = 0
-          while (i < batchSize && prev.hasNext) {
-            val sample = prev.next()
-            val featureLength = sample.getFeature().nElement()
-            val labelLength = sample.getLabel().nElement()
-            if (featureSize == null || labelSize == null
-              || _featureLength != featureLength
-              || _labelLength != labelLength) {
-              featureSize = Array(1) ++ sample.getFeature().size
-              labelSize = Array(1) ++ sample.getLabel().size
-              _featureLength = featureLength
-              _labelLength = labelLength
+          if (isTable) {
+            var i = 0
+            while (i < batchSize && prev.hasNext) {
+              val sample = prev.next()
+              /**
+              * Each Tensor in the Table should be independent
+              * Thus the resizeAs function will only affect one sample
+              */
+              featureArrayOfTensor(i).resizeAs(sample.feature()).copy(sample.feature())
+              labelArrayOfTensor(i).resizeAs(sample.label()).copy(sample.label())
+              featureTable(i + 1) = featureArrayOfTensor(i)
+              labelTable(i + 1) = labelArrayOfTensor(i)
+              i += 1
             }
-            if (featureData == null || featureData.length < batchSize * featureLength) {
-              featureData = new Array[T](batchSize * featureLength)
+            var j = i + 1
+            while (j <= featureTable.length) {
+              featureTable.remove(j)
+              j += 1
             }
-            if (labelData == null || labelData.length < batchSize * labelLength) {
-              labelData = new Array[T](batchSize * labelLength)
+            j = i + 1
+            while (j <= labelTable.length) {
+              labelTable.remove(j)
```

Comment:
Table.remove() is very costly; we should introduce something like reset or create a new Table